package com.rvt.bfcrmod.events;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.UUID;

import com.rvt.bfcrmod.BetterForgeChat;
import com.rvt.bfcrmod.MarkdownFormatter;
import com.rvt.bfcrmod.TextFormatter;
import com.rvt.bfcrmod.config.ConfigHandler;
import com.rvt.bfcrmod.config.IReloadable;
import com.rvt.bfcrmod.config.PermissionsHandler;
import com.rvt.bfcrmod.utils.BetterForgeChatUtilities;
import com.mojang.authlib.GameProfile;

import net.minecraft.ChatFormatting;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.MutableComponent;
import net.minecraft.network.chat.Style;
import net.minecraft.network.chat.contents.TranslatableContents;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.player.Player;
import net.neoforged.neoforge.event.ServerChatEvent;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;

@EventBusSubscriber
public class ChatEventHandler implements IReloadable {
	private static SimpleDateFormat timestampFormat = null;
	private static boolean markdownEnabled = false;
	private static String chatMessageFormat = "";
	private static String chatMessageColor = "";
	private static boolean loaded = false;
	
        @Override
	public void reloadConfigOptions() {
		loaded = false;
		timestampFormat = ConfigHandler.config.enableTimestamp.get() ? new SimpleDateFormat(ConfigHandler.config.timestampFormat.get()) : null;
		markdownEnabled = ConfigHandler.config.enableMarkdown.get();
		chatMessageFormat = ConfigHandler.config.chatMessageFormat.get();
		chatMessageColor = ConfigHandler.config.chatMessageColor.get();
		loaded = true;
	}

	public static String getChatMessageColor() {
		BetterForgeChat.LOGGER.debug("Chat color from config: {}",chatMessageColor);
		return chatMessageColor;
	}

	public static Style getHoverClickEventStyle(Component old) {
		if(old instanceof TranslatableContents tcmp) {
            Object[] args = tcmp.getArgs();
			for(Object arg : args) {
				if(arg instanceof MutableComponent tc) {
                    if(tc.getStyle() != null && tc.getStyle().getClickEvent() != null)
						return ((MutableComponent) arg).getStyle();
				}
			}
		}
		return null;
	}
	
	@SubscribeEvent
    public static void onServerChat(ServerChatEvent e) {
		if(!loaded) return; // Just do nothing until everything's ready to go!
    	ServerPlayer player = e.getPlayer();
        UUID uuid = player.getUUID();
        String msg = e.getMessage().getString();
		if(msg.isEmpty()) return;
		String tstamp = timestampFormat == null ? "" : timestampFormat.format(new Date());
		String name = BetterForgeChatUtilities.getRawPreferredPlayerName(player);
		BetterForgeChat.LOGGER.debug("global message format: {}",chatMessageFormat);
		String fmat = chatMessageFormat.replace("$time", tstamp).replace("$name", name);
		BetterForgeChat.LOGGER.debug("formatted: {}",fmat);
		MutableComponent beforeMsg = TextFormatter.stringToFormattedText(player, fmat.substring(0, fmat.indexOf("$msg")));
		BetterForgeChat.LOGGER.debug("before message: {}",beforeMsg.toString());
		MutableComponent afterMsg = TextFormatter.stringToFormattedText(player, fmat.substring(fmat.indexOf("$msg") + 4));
		BetterForgeChat.LOGGER.debug("after message: {}",afterMsg.toString());
		boolean enableColor = PermissionsHandler.playerHasPermission(uuid, PermissionsHandler.coloredChatNode);
		boolean enableStyle = PermissionsHandler.playerHasPermission(uuid, PermissionsHandler.styledChatNode);

		// Create an error message if the player isn't allowed to use styles/colors
		String emsg = "";
		if(!enableColor && TextFormatter.messageContainsColorsOrStyles(msg, true))
			emsg = "You are not permitted to use colors";
		if(!enableStyle && TextFormatter.messageContainsColorsOrStyles(msg, false))
			emsg += !emsg.isEmpty() ? " or styles" : "You are not permitted to use styles";
		if(!emsg.isEmpty()) {
			MutableComponent ecmp = Component.literal(emsg + "!");
			ecmp.withStyle(ChatFormatting.BOLD);
			ecmp.withStyle(ChatFormatting.RED);
			player.sendSystemMessage(ecmp);
		}
		// Convert markdown to normal essentials formatting
		if(markdownEnabled && enableStyle && PermissionsHandler.playerHasPermission(uuid, PermissionsHandler.markdownChatNode)) {
			BetterForgeChat.LOGGER.debug("pre markdown formatted text: {}", msg);
			msg = MarkdownFormatter.markdownStringToFormattedString(msg);
			BetterForgeChat.LOGGER.debug("post markdown formatted text: {}", msg);
		}

		// Start generating the main TextComponent
		MutableComponent msgComp = TextFormatter.stringToFormattedText(player, msg, enableColor, enableStyle);

		// Append the hover and click event crap
		Style sty = getHoverClickEventStyle(e.getMessage());
		MutableComponent ecmp = Component.empty();
		if(sty != null && sty.getHoverEvent() != null)
			ecmp.setStyle(sty);
//		e.setCanceled(true);

		MutableComponent newMessage = msgComp.append(afterMsg);

        e.setMessage(newMessage);
//		player.server.execute(() -> {
//			BetterForgeChat.LOGGER.info("[CHAT] "+newMessage.getString());
//			ServerMessageEvent.broadcastMessage(player.level(), newMessage);
//		});

    }
}